torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -370,7 +370,15 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
             work.type_,
             work.id_);
         if (message.isRequest()) {
-          send(work.from_, cb_->operator()(message));
+          auto futureResponse = cb_->operator()(message);
+          if (futureResponse->completed()) {
+            send(work.from_, std::move(*futureResponse).moveMessage());
+          } else {
+            auto fromId = work.from_.id_;
+            futureResponse->addCallback([this, fromId](const Message& msg) {
+              send(getWorkerInfo(fromId), std::move(const_cast<Message&>(msg)));
```

Comment:
That's the issue with the FutureMessage api as-is - we get(const Message&), though we know that we're the only (or last) one using that FutureMessage, since we don't expose it externally.

My feeling is that changing to non-const in the api is probably worse here, since we allow a list of callbacks.

Reply:
ah, nevermind - I just updated to remove the const_cast<>.

(I was originally concerned with ref-cycle loops if I placed a reference to the future in the callback captures, but given that they're cleared afterwards, it's not an issue.)

Reply:
Thanks,  @jjlilley!